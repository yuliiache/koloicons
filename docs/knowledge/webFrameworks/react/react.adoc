== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







