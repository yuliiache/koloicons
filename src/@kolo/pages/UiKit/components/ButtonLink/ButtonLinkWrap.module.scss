.buttonLinkWrap {
  margin-left: 20px;
}
