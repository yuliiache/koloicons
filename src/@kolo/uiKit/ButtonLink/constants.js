export const BUTTONLINK_TEST_ID = 'buttonLinkComponent';

