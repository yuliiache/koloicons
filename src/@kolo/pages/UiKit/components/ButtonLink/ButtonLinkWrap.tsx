import ButtonLink from '@kolo/uiKit/ButtonLink/ButtonLink';
import { AppRoute } from 'constants/AppRoute';
import {FC} from 'react';

import styles from './ButtonLinkWrap.module.scss';

export const ButtonLinkWrap: FC = () => {
  return (
    <div className={styles.buttonLinkWrap}>
      <ButtonLink
        text="Button link"
        url={AppRoute.HOME}
      />
    </div>
  );
};
