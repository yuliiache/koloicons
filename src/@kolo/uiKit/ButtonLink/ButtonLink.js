import React from 'react';
import {Link as InternalLink} from 'react-router-dom';
import PropTypes from 'prop-types';

import {Button} from '../Button/Button';
import {BUTTON_COLOR, BUTTON_SIZE, BUTTON_TYPE} from '../Button/constants';

import {BUTTONLINK_TEST_ID} from './constants';

export const ButtonLink = ({
  color = BUTTON_COLOR.PRIMARY,
  size = BUTTON_SIZE.SMALL,
  text,
  isDisabled,
  type = BUTTON_TYPE.BUTTON,
  url,
}) => {
  const isInternalUrl = url.startsWith('/');

  if (isInternalUrl) {
    return (
      <InternalLink
        to={url}
        data-testid={BUTTONLINK_TEST_ID}
      >
        <Button
          color={color}
          size={size}
          text={text}
          disabled={isDisabled}
          type={type}
        />
      </InternalLink>
    );
  }

  return (
    <a
      href={url}
      target="_blank"
      rel="noopener noreferrer nofollow"
      data-testid={BUTTONLINK_TEST_ID}
    >
      <Button
        color={color}
        size={size}
        text={text}
        disabled={isDisabled}
        type={type}
      />
    </a>
  );
};

ButtonLink.propTypes = {
  size: PropTypes.oneOf(Object.values(BUTTON_SIZE)),
  color: PropTypes.oneOf(Object.values(BUTTON_COLOR)),
  text: PropTypes.string.isRequired,
  isDisabled: PropTypes.bool,
  type: PropTypes.oneOf(Object.values(BUTTON_TYPE)),
  url: PropTypes.string.isRequired,
};
