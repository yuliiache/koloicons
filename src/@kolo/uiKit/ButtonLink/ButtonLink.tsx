import {FC} from 'react';
import {Link as InternalLink} from 'react-router-dom';

import Button from '../Button/Button';
import {ButtonProps} from '../Button/Button';
import {ButtonColor, ButtonSize} from '../Button/constants';
import {BUTTONLINK_TEST_ID} from './constants';

interface ButtonLinkProps extends ButtonProps {
  url: string;
}

const ButtonLink: FC<ButtonLinkProps> = ({
  url,
  text,
  isDisabled,
  children,
  color = ButtonColor.PRIMARY,
  size = ButtonSize.SMALL,
}) => {
  const isInternalUrl: boolean = url.startsWith('/');

  if (isInternalUrl) {
    return (
      <InternalLink
        to={url}
        data-testid={BUTTONLINK_TEST_ID}
      >
        <Button
          text={text}
          isDisabled={isDisabled}
          color={color}
          size={size}
        >
          {children}
        </Button>
      </InternalLink>
    );
  }

  return (
    <a
      href={url}
      target="_blank"
      rel="noopener noreferrer nofollow"
      data-testid={BUTTONLINK_TEST_ID}
    >
      <Button
        text={text}
        isDisabled={isDisabled}
        color={color}
        size={size}
      >
        {children}
      </Button>
    </a>
  );
};
export default ButtonLink;
