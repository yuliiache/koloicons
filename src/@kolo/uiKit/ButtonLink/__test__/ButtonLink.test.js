import {BrowserRouter as Router} from 'react-router-dom';
import {cleanup, render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ButtonLink from '../ButtonLink';

describe('ButtonLink component', () => {
  afterEach(cleanup);

  const commonProps = {
    text: 'Link',
    isDisabled: false,
    url: '/',
  };

  describe('snapshot', () => {
    it('should render button link', () => {
      const {asFragment} = render(
        <Router>
          <ButtonLink {...commonProps} />
        </Router>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should have text', () => {
    const buttonText = 'Text';
    const {getByText} = render(
      <Router>
        <ButtonLink
          {...commonProps}
          text={buttonText}
        />
      </Router>
    );
    const buttonTextElement = getByText(buttonText);

    expect(buttonTextElement).toBeInTheDocument();
  });

  it('should have href attribute', () => {
    render(
      <Router>
        <ButtonLink {...commonProps} />
      </Router>
    );

    const linkElement = screen.getByRole('link');

    expect(linkElement).toHaveAttribute('href', `${commonProps.url}`);
  });

  it('should navigate when clicked', () => {
    render(
      <Router>
        <ButtonLink {...commonProps} />
      </Router>
    );

    const linkElement = screen.getByRole('link');

    expect(linkElement).toBeInTheDocument();

    userEvent.click(linkElement);

    expect(window.location.pathname).toBe(`${commonProps.url}`);
  });
});
