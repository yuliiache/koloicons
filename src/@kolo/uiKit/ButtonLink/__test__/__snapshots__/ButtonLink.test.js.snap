// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonLink component snapshot should render button link 1`] = `
<DocumentFragment>
  <a
    data-testid="buttonLinkComponent"
    href="/"
  >
    <button
      class="button button-small button-primary"
      data-testid="buttonComponent"
      type="button"
    >
      <span
        data-testid="buttonComponentText"
      >
        Link
      </span>
    </button>
  </a>
</DocumentFragment>
`;
